File: src/leetcode_grind/day_314.rs
// https://leetcode.com/problems/is-subsequence/editorial
pub fn is_subsequence_two_pointers(s: String, t: String) -> bool {
    let s = s.chars().collect::<Vec<_>>();
    let t = t.chars().collect::<Vec<_>>();
    let mut left = 0;
    let mut right = 0;
    while left < s.len() && right < t.len() {
        if s[left] == t[right] {
            left += 1;
        }
        right += 1;
    }
    left == s.len()
}

pub fn is_subsequence_dp_lcs_i(s: String, t: String) -> bool {
    let s = s.chars().collect::<Vec<_>>();
    let t = t.chars().collect::<Vec<_>>();

    let (m, n) = (s.len(), t.len());
    let mut dp = vec![vec![0; n + 1]; m + 1];

    for row in 1..=m {
        for col in 1..=n {
            if s[row - 1] == t[col - 1] {
                dp[row][col] = dp[row - 1][col - 1] + 1;
            } else {
                dp[row][col] = dp[row][col - 1].max(dp[row - 1][col]);
            }
        }
    }

    for &val in &dp[m] {
        if val == s.len() {
            return true;
        }
    }

    false
}

pub fn is_subsequence_dp_lcs_ii(s: String, t: String) -> bool {
    let s = s.chars().collect::<Vec<_>>();
    let t = t.chars().collect::<Vec<_>>();

    let (m, n) = (s.len(), t.len());

    if m == 0 {
        return true;
    }

    let mut dp = vec![vec![0; n + 1]; m + 1];
    for col in 1..=n {
        for row in 1..=m {
            if s[row - 1] == t[col - 1] {
                dp[row][col] = dp[row - 1][col - 1] + 1;
            } else {
                dp[row][col] = dp[row - 1][col].max(dp[row][col - 1]);
            }
        }
        if dp[m][col] == m {
            return true;
        }
    }
    false
}

// https://leetcode.com/problems/shortest-way-to-form-string/description/
pub fn shortest_way_bin_search(source: String, target: String) -> i32 {
    fn is_subsequece(s: &Vec<char>, t: &Vec<char>) -> bool {
        let mut i = 0;
        for &ch in t {
            if ch == s[i] {
                i += 1;
            }
            if i == s.len() {
                return true;
            }
        }
        return false;
    }

    let (s, t) = (
        source.chars().collect::<Vec<_>>(),
        target.chars().collect::<Vec<_>>(),
    );

    let mut left = 0;
    let mut right = target.len() as i32 + 1;

    while left < right {
        let mid = (left + right) / 2;
        if is_subsequece(&t, &s.repeat(mid as usize)) {
            right = mid;
        } else {
            left = mid + 1;
        }
    }

    if left > t.len() as i32 {
        return -1;
    }
    left
}

pub fn shortest_way_2d_array(source: String, target: String) -> i32 {
    let source = source.chars().collect::<Vec<_>>();
    let target = target.chars().collect::<Vec<_>>();

    let mut next_occurence = vec![vec![-1; 26]; source.len()];
    next_occurence[source.len() - 1][source[source.len() - 1] as usize - 'a' as usize] =
        source.len() as i32 - 1;

    for idx in (0..source.len() - 1).rev() {
        for c in 0..26 {
            next_occurence[idx][c] = next_occurence[idx + 1][c];
        }
        next_occurence[idx][source[idx] as usize - 'a' as usize] = idx as i32;
    }

    let mut source_iterator = 0;
    let mut count = 1;

    for c in target {
        if next_occurence[0][c as usize - 'a' as usize] == -1 {
            return -1;
        }

        if source_iterator == source.len()
            || next_occurence[source_iterator][c as usize - 'a' as usize] == -1
        {
            count += 1;
            source_iterator = 0;
        }

        source_iterator = next_occurence[source_iterator][c as usize - 'a' as usize] as usize + 1;
    }

    count
}

File: java/leetcode_grind/Day314.java
package leetcode_grind;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Day314 {
    // https://leetcode.com/problems/median-of-a-row-wise-sorted-matrix/description/
    static class Solution1 {
        public int matrixMedian(int[][] grid) {
            var binSearch = new Object() {
                int apply(int[] cols, int target) {
                    var left = 0;
                    var right = cols.length - 1;

                    while (left <= right) {
                        var mid = (left + right) / 2;
                        if (cols[mid] >= target) {
                            right = mid - 1;
                        } else {
                            left = mid + 1;
                        }
                    }
                    return left;
                }
            };

            var m = grid.length;
            var n = grid[0].length;

            var half = (m * n) / 2;
            var left = 1;
            var right = (int) 10e6;

            while (left + 1 < right) {
                var mid = (left + right) / 2;

                var sum = 0;
                for (var col : grid) {
                    sum += binSearch.apply(col, mid);
                }

                if (sum > half) {
                    right = mid;
                } else {
                    left = mid;
                }
            }

            return left;
        }
    }

    // https://leetcode.com/problems/is-subsequence/description
    static class Solution2 {
        public boolean isSubsequence(String s, String t) {
            var indices = new HashMap<Character, List<Integer>>();
            var i = 0;
            for (char c : t.toCharArray()) {
                indices.putIfAbsent(c, new ArrayList<>());
                indices.get(c).add(i++);
            }

            var currMatchIndex = -1;
            for (char c : s.toCharArray()) {
                if (!indices.containsKey(c)) {
                    return false;
                }

                var isMatched = false;
                for (var matchIndex : indices.get(c)) {
                    if (currMatchIndex < matchIndex) {
                        currMatchIndex = matchIndex;
                        isMatched = true;
                        break;
                    }
                }

                if (!isMatched) {
                    return false;
                }
            }

            return true;
        }
    }

    // https://leetcode.com/problems/number-of-matching-subsequences/description/
    static class Solution3 {
        public int numMatchingSubseqHashMapTLE(String s, String[] words) {
            var indices = new HashMap<Character, List<Integer>>();
            var i = 0;
            for (var ch : s.toCharArray()) {
                indices.putIfAbsent(ch, new ArrayList<>());
                indices.get(ch).add(i);
                i++;
            }

            Function<String, Boolean> check = (str) -> {
                var last = -1;
                for (var ch : str.toCharArray()) {
                    if (!indices.containsKey(ch)) {
                        return false;
                    }
                    var found = false;

                    for (var idx : indices.get(ch)) {
                        if (idx > last) {
                            last = idx;
                            found = true;
                            break;
                        }
                    }

                    if (!found) {
                        return false;
                    }
                }

                return true;
            };

            var ans = 0;
            for (var w : words) {
                if (check.apply(w)) {
                    ans++;
                }
            }
            return ans;
        }

        public int numMatchingSubseqBucketsApproach(String s, String[] words) {
            class Node {
                String word;
                int index;

                Node(String w, int i) {
                    word = w;
                    index = i;
                }
            }

            var ans = 0;
            @SuppressWarnings("unchecked")
            ArrayList<Node>[] heads = new ArrayList[26];
            for (var i = 0; i < 26; ++i) {
                heads[i] = new ArrayList<Node>();
            }

            for (var word : words) {
                heads[word.charAt(0) - 'a'].add(new Node(word, 0));
            }

            for (var c : s.toCharArray()) {
                var oldBucket = heads[c - 'a'];
                heads[c - 'a'] = new ArrayList<>();

                for (var node : oldBucket) {
                    node.index++;
                    if (node.index == node.word.length()) {
                        ans++;
                    } else {
                        heads[node.word.charAt(node.index) - 'a'].add(node);
                    }
                }
            }

            return ans;
        }
    }

    // https://leetcode.com/problems/shortest-way-to-form-string/description/
    static class Solution4 {
        public int shortestWay(String source, String target) {
            var invertedIndex = new HashMap<Character, List<Integer>>();
            var i = 0;
            for (var ch : source.toCharArray()) {
                invertedIndex.putIfAbsent(ch, new ArrayList<>());
                invertedIndex.get(ch).add(i);
                i++;
            }

            for (var ch : target.toCharArray()) {
                if (!invertedIndex.containsKey(ch)) {
                    return -1;
                }
            }

            var sourceIterator = 0;
            var count = 1;

            for (var ch : target.toCharArray()) {
                var indices = invertedIndex.get(ch);
                var index = Collections.binarySearch(indices, sourceIterator);

                if (index < 0) {
                    index = -index - 1;
                }

                if (index == indices.size()) {
                    count++;
                    sourceIterator = indices.get(0) + 1;
                } else {
                    sourceIterator = indices.get(index) + 1;
                }
            }

            return count;
        }
    }
}
